src/main/asciidoc/futures.adoc

```diff
@@ -0,0 +1,155 @@
+== Future results
+
+Vert.x 4 use futures to represent asynchronous results.
+
+Any asynchronous method returns a {@link io.vertx.core.Future} object for the result of the call:
+a _success_ or a _failure_.
+
+You cannot interact directly with the result of a future, instead you need to set a handler that will be called when the future completes and the result is available, like any other kind of event.
+
+[source,$lang]
+----
+{@link examples.CoreExamples#exampleFuture1}
+----
+
+NOTE: Vert.x 3 provides a callback-only model.
+To allow an easy migration to Vert.x 4 we decided that each asynchronous method has also a callback version.
+The `props` method above has also a version {@link io.vertx.core.file.FileSystem#props(java.lang.String, io.vertx.core.Handler)} with a callback as method argument.
+
+[CAUTION]
+====
+Do not confuse _futures_ with _promises_.
+
+If futures represent the "read-side" of an asynchronous result, promises are the "write-side".
+They allow you to defer the action of providing a result.
+
+In most cases, you don't need to create promises yourself in a Vert.x application.
+<<_future_composition>> and <<_future_coordination>> provide you with the tools to transform and merge asynchronous results.
+
+However, if, in your codebase, you have legacy methods which use callbacks, you can leverage the fact that a promise extends {@link io.vertx.core.Handler io.vertx.core.Handler<io.vertx.core.AsyncResult>}:
+
+[source,$lang]
+----
+{@link examples.CoreExamples#promiseAsHandler}
+----
+====
+
+[CAUTION]
+====
+Terminal operations like `onSuccess`, `onFailure` and `onComplete` provide no guarantee whatsoever regarding the invocation order of callbacks.
+
+Consider a future on which 2 callbacks are registered:
+
+[source,$lang]
+----
+{@link examples.CoreExamples#promiseCallbackOrder}
+----
+
+It is possible that the second callback is invoked before the first one.
```

Comment:
I think here we should mention that ordered side effect behaviour can be achieved with `andThen`